# 🏒  NHL Line Combination Optimizer
### Top Line Combinations of the 2025-26 Playoff Odds Leaders 
(Data and Predications from: MoneyPuck on March 16, 2026)

In [26]:
TOP_TEAMS = ['COL', 'DAL', 'MIN', 'CAR', 'BUF', 'TBL', 'ANA', 'VGK']

lines_2023 = pd.read_csv("lines_2023.csv")
lines_2024 = pd.read_csv("lines_2024.csv")

full_df = pd.concat([lines_2023, lines_2024], axis=0).reset_index(drop=True)

full_df.shape

(6004, 108)

After loading and merging our databases, it looks like we have 6,004 pairings and 108 columns! Let's take a look at all of the possible columns we can use later to pick out the best line pairings.

In [12]:
full_df.columns.to_list()

['lineId',
 'season',
 'name',
 'team',
 'position',
 'situation',
 'games_played',
 'icetime',
 'iceTimeRank',
 'xGoalsPercentage',
 'corsiPercentage',
 'fenwickPercentage',
 'xOnGoalFor',
 'xGoalsFor',
 'xReboundsFor',
 'xFreezeFor',
 'xPlayStoppedFor',
 'xPlayContinuedInZoneFor',
 'xPlayContinuedOutsideZoneFor',
 'flurryAdjustedxGoalsFor',
 'scoreVenueAdjustedxGoalsFor',
 'flurryScoreVenueAdjustedxGoalsFor',
 'shotsOnGoalFor',
 'missedShotsFor',
 'blockedShotAttemptsFor',
 'shotAttemptsFor',
 'goalsFor',
 'reboundsFor',
 'reboundGoalsFor',
 'freezeFor',
 'playStoppedFor',
 'playContinuedInZoneFor',
 'playContinuedOutsideZoneFor',
 'savedShotsOnGoalFor',
 'savedUnblockedShotAttemptsFor',
 'penaltiesFor',
 'penalityMinutesFor',
 'faceOffsWonFor',
 'hitsFor',
 'takeawaysFor',
 'giveawaysFor',
 'lowDangerShotsFor',
 'mediumDangerShotsFor',
 'highDangerShotsFor',
 'lowDangerxGoalsFor',
 'mediumDangerxGoalsFor',
 'highDangerxGoalsFor',
 'lowDangerGoalsFor',
 'mediumDangerGoalsFor',
 'high

In [16]:
print(full_df["situation"].value_counts())
print(full_df["position"].value_counts())

situation
5on5    6004
Name: count, dtype: int64
position
line       4004
pairing    2000
Name: count, dtype: int64


The database focuses around lines/pairings happening on regular play situations (5on5) with two sets of positions, forwards and defensemen (line, pairing)

Let's now take a look at the teams representated in this database, making sure everything is correct.

In [20]:
full_df['team'].unique()

array(['FLA', 'DAL', 'SJS', 'STL', 'TOR', 'WSH', 'NYR', 'PHI', 'CGY',
       'LAK', 'VAN', 'SEA', 'MIN', 'CBJ', 'NYI', 'ARI', 'BUF', 'TBL',
       'ANA', 'OTT', 'CHI', 'EDM', 'MTL', 'VGK', 'DET', 'CAR', 'NSH',
       'BOS', 'PIT', 'COL', 'WPG', 'NJD', 'UTA'], dtype=object)

Looks like we need to account for the relocation of the Arizona Coyotes to Utah.

In [21]:
full_df['team'] = full_df['team'].replace('ARI', 'UTA')

In [22]:
full_df['team'].nunique()

32

Now, we need to filter the database down to the eight top teams

In [27]:
df_top = full_df[full_df['team'].isin(TOP_TEAMS)]

In [30]:
print(df_top['team'].unique())
print(df_top.shape)

['DAL' 'MIN' 'BUF' 'TBL' 'ANA' 'VGK' 'CAR' 'COL']
(1482, 108)
